### ISSUE TITLE
Home-assistant doesn't build on unstable

### ISSUE THREAD

**Issue by netixx** [NixOS/nixpkgs#55527]
## Issue description
Home assistant package, version 0.86.4 does not build in unstable channel.

It seems it tries to build a wheel file (command `bdist_wheel`), but fails because python-wheel is probably missing.

### Steps to reproduce
Use `services.home-assistant.package = unstable.home-assistant` in nixos config.

## Technical details
```
building '/nix/store/w9kg3n0yyfdrrksrd1iis5s317gs1xp4-homeassistant-0.86.4.drv'...
unpacking sources
unpacking source archive /nix/store/pxv5z4gm2wc4rwza3rnvpsw0lcn218bi-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/virtualization/vagrant/provision.sh
patching sources
configuring
building
usage: nix_run_setup [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: nix_run_setup --help [cmd1 cmd2 ...]
   or: nix_run_setup --help-commands
   or: nix_run_setup cmd --help

error: invalid command 'bdist_wheel'
builder for '/nix/store/w9kg3n0yyfdrrksrd1iis5s317gs1xp4-homeassistant-0.86.4.drv' failed with exit code 1
cannot build derivation '/nix/store/zh1vshn4pr0pgkdvxn1vx297wghwxh28-unit-home-assistant.service.drv': 1 dependencies couldn't be built
```

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.95, NixOS, 18.09pre-git (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(root): `"nixos-18.09.2127.5225d4bf019, unstable-19.03pre168613.ffc604e5579"`
 - nixpkgs: `/var/nixpkgs`

**Comment by dotlambda** [comment]
I don't see any failure at https://hydra.nixos.org/job/nixpkgs/trunk/home-assistant.x86_64-linux or https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.home-assistant.x86_64-linux.
Can you please tell me exactly what `unstable.home-assistant` refers to so I can reproduce it?

Next time, please ping the maintainer.

**Comment by dotlambda** [comment]
Actually, using `nix build -f https://github.com/NixOS/nixpkgs/archive/ffc604e5579.tar.gz home-assistant` (commit ffc604e5579 is apparently what your unstable refers to) yields a different package (`/nix/store/iqgjd1b3mlk5fgd73m1zn9a6kx5y7y9q-homeassistant-0.86.4`).
Your configuration of the Home Assistant service might add some additional dependencies which cause the build failure. You might want to paste that configuration or, even better, find out which packages are added to `extraComponentPackages`.

**Comment by netixx** [comment]
Thank you for your answer ! Sorry for not mentioning you when I opened the issue (I sort of thought it was automatic)...

I do have a specific configuration because my `lifx` lights were not detected because of an import problem (which reminds me I should have opened an issue/fix at the time...). I didn't think this was causing the problem because I didn't see it in the stack trace...


For the record, I had the following stack trace with the lifx component:
```
 Traceback (most recent call last)
   File "/nix/store/mkgffgcfk63fkrqz1f7pnfyms1ffq91j-homeassistant-0.86.4/lib/python3.7/site-packages/homeassistant/helpers/entity_platform.py", line 128, in _async_setup_platform
     SLOW_SETUP_MAX_WAIT, loop=hass.loop)
   File "/nix/store/ydk0mfpvn9smcmn72wc9i20slv1d2b79-python3-3.7.2/lib/python3.7/asyncio/tasks.py", line 416, in wait_for
     return fut.result()
   File "/nix/store/mkgffgcfk63fkrqz1f7pnfyms1ffq91j-homeassistant-0.86.4/lib/python3.7/site-packages/homeassistant/components/light/lifx.py", line 157, in async_setup_entry
     lifx_manager = LIFXManager(hass, async_add_entities)
   File "/nix/store/mkgffgcfk63fkrqz1f7pnfyms1ffq91j-homeassistant-0.86.4/lib/python3.7/site-packages/homeassistant/components/light/lifx.py", line 211, in __init__
     self.effects_conductor = aiolifx_effects().Conductor(loop=hass.loop)
   File "/nix/store/mkgffgcfk63fkrqz1f7pnfyms1ffq91j-homeassistant-0.86.4/lib/python3.7/site-packages/homeassistant/components/light/lifx.py", line 129, in aiolifx_effects
     import aiolifx_effects as aiolifx_effects_module
 ModuleNotFoundError
```

I use the following configuration to build the home-assistant (where `aiolifx/default.nix` and `aiolifx_effects/default.nix` are python packages I will try to submit upstream - both build correctly):
```nix
{config, pkgs, ...}:
with pkgs.python3Packages;
let
unstable = import <unstable> {};
python = unstable.python3.override {
    packageOverrides = self: super: rec {
      aiolifx = super.callPackage ./hass/lifx/aiolifx/default.nix { };
      aiolifx_effects = super.callPackage ./hass/lifx/aiolifx_effects/default.nix { };
    };
  };
in
{
      services.home-assistant = {
        enable = true;
        # package = unstable.home-assistant;
        package = unstable.home-assistant.override {
          extraPackages = ps: with ps; [
            xmltodict
            distro
            python.pkgs."aiolifx_effects"
            # pyopenuv
          ];
        };
        autoExtraComponents = true;
    };
}
```
Unstable is the current unstable channel (`nix-channel --add unstable https://nixos.org/channels/nixos-unstable unstable`).

If I switch back to original unstable.home-assistant, it works.

Am I missing something ?

**Comment by dotlambda** [comment]
I think the problem is that `python` is different from the `home-assistant` package is using. It's even a different version (3.6 vs 3.7).
Try the following:
```
package = unstable.home-assistant.override {
  extraPackages = ps: let
    aiolifx = ps.callPackage ./hass/lifx/aiolifx/default.nix { };
    aiolifx_effects = ps.callPackage ./hass/lifx/aiolifx_effects/default.nix {
      inherit aiolifx;
    };
  in with ps; [
    xmltodict
    distro
    aiolifx_effects
  ];
};
```

**Comment by dotlambda** [comment]
Actually, the best way to do this is
```
package = unstable.home-assistant.override {
  packageOverrides = self: super: {
    aiolifx = self.callPackage ./hass/lifx/aiolifx/default.nix { };
    aiolifx_effects = self.callPackage ./hass/lifx/aiolifx_effects/default.nix { };
  };
  extraPackages = ps: with ps; [
    xmltodict
    distro
    aiolifx_effects
  ];
};
```

Since
```
nix-build -I nixpks=https://github.com/NixOS/nixpkgs/archive/ffc604e5579.tar.gz -E "with import ./. {}; home-assistant.override { extraPackages = ps: with ps; [ distro xmltodict ]; }"
```
builds just fine, the only thing I can recommend doing is setting `autoExtraComponents = false` and removing the `extraPackages` one by one to nail down which dependency is causing the build failure.

**Comment by netixx** [comment]
THANK YOU!

I tried the last version you suggested and it worked 😄 !

The thing I had before always seemed awkward to me (but it used to work with previous hass versions...). It is difficult to find up-to-date information regarding python and nix because there has been a lot of evolutions (.override, overlays...).

Should I try to submit a PR to add those package to the `lifx` component dependencies and nixpkgs ?

**Comment by dotlambda** [comment]
> I tried the last version you suggested and it worked  !

I'm glad I was able to help.

> It is difficult to find up-to-date information regarding python and nix because there has been a lot of evolutions (.override, overlays...).

Your previous usage of `packageOverrides` was almost correct. You would have had to add `home-assistant.override { python3 = python; }` to make it perfect. However, that would not have worked because of https://github.com/NixOS/nixpkgs/issues/44426. That's why I added the `home-assistant`-specific `packageOverrides` argument in https://github.com/NixOS/nixpkgs/commit/33be3c463022748449c46e4027d23ceb3d74671c#diff-7ce5df58d1f5c8bb8215b0f75c81b8a0R9. It won't work for other packages.

> Should I try to submit a PR to add those package to the `lifx` component dependencies and nixpkgs ?

I'd warmly welcome such a PR :)
